search-provider/eks-search-app.c

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2016 Endless Mobile, Inc. */
+
+#include "eks-search-app.h"
+
+#include "eks-search-provider.h"
+#include "eks-search-provider-dbus.h"
+#include "eks-subtree-dispatcher.h"
+
+#include <systemd/sd-bus.h>
+
+/**
+ * EksSearchProvider:
+ *
+ * A search provider for a single knowledge app, to be used through dbus by the
+ * shell's global search. Requires the app id of the knowledge app it should run
+ * searches for.
+ *
+ * This search provider will activate the actual knowledge app over dbus with a
+ * result or search to display.
+ */
+struct _EksSearchApp
+{
+  GApplication parent_instance;
+
+  EksSubtreeDispatcher *dispatcher;
+  // Hash table with app id string keys, EksSearchProvider values
+  GHashTable *app_search_providers;
+};
+
+G_DEFINE_TYPE (EksSearchApp,
+               eks_search_app,
+               G_TYPE_APPLICATION)
+
+static void
+eks_search_app_finalize (GObject *object)
+{
+  EksSearchApp *self = EKS_SEARCH_APP (object);
+
+  g_clear_object (&self->dispatcher);
+  g_clear_pointer (&self->app_search_providers, g_hash_table_unref);
+
+  G_OBJECT_CLASS (eks_search_app_parent_class)->finalize (object);
+}
+
+static gboolean
+eks_search_app_register (GApplication *application,
+                         GDBusConnection *connection,
+                         const gchar *object_path,
+                         GError **error)
+{
+  EksSearchApp *self = EKS_SEARCH_APP (application);
+
+  eks_subtree_dispatcher_register (self->dispatcher, connection, object_path);
```

Comment:
I would suggest amending the API of `eks_subtree_dispatcher_register()` to have an error-out parameter, since this vfunc has an error-out parameter itself.